### ISSUE TITLE
CSRF Issue in Version 2.25 with Hazelcast session replication

### ISSUE THREAD

**Issue by imilic1** [jhipster/generator-jhipster#2485]
Hi,

First of all thanks to all contributors for this amazing framework. I am using some nice tricks from jHipster for my project and after upgrading to version 2.25, I hit the issue with CSRF token.
This issue exists also when specific configuration is created from scratch.

Steps to Replicate on clean installation:

1. Create new jHipster Application from scratch after upgrading Jhipster generator to version 2.25. Here are exact configuration options:

Welcome to the JHipster Generator v2.25.0

? (1/15) What is the base name of your application? myapp
? (2/15) What is your default Java package name? com.myapps.system
? (3/15) Which *type* of authentication would you like to use? HTTP Session Authentication (stateful, default Spring Security mechanism)
? (4/15) Which *type* of database would you like to use? SQL (H2, MySQL, PostgreSQL, Oracle)
? (5/15) Which *production* database would you like to use? PostgreSQL
? (6/15) Which *development* database would you like to use? PostgreSQL
? (7/15) Do you want to use Hibernate 2nd level cache? Yes, with HazelCast (distributed cache, for multiple nodes)
? (8/15) Do you want to use a search engine in your application? Yes, with ElasticSearch
? (9/15) Do you want to use clustered HTTP sessions? Yes, with HazelCast
? (10/15) Do you want to use WebSockets? Yes, with Spring Websocket
? (11/15) Would you like to use Maven or Gradle for building the backend? Maven
? (12/15) Would you like to use Grunt or Gulp.js for building the frontend? Grunt (recommended)
? (13/15) Would you like to use the LibSass stylesheet preprocessor for your CSS? Yes
? (14/15) Would you like to enable translation support with Angular Translate? Yes
? (15/15) Which testing frameworks would you like to use? Gatling

2. Create DB, add user/pass to application-dev.yaml
3. Run mvn spring-boot:run from via cmd to kick off the application.
4. Once Started, log into application with admin/admin
5. Navigate to User Management Application and update users last name for example to anything.
6. Press Save button.
ISSUE: The change doesn't get saved and you get the following server response:

{"timestamp":"2015-12-14T22:38:47.320+0000","status":403,"error":"Forbidden","message":"Expected CSRF token not found. Has your session expired?","path":"/api/users"}

Press the save again.
ISSUE: Again the change doesn't get saved and now you get the following server response:

{"timestamp":"2015-12-14T22:38:54.575+0000","status":403,"error":"Forbidden","message":"Invalid CSRF Token '67501fe2-c436-477c-a2ec-bc9827d805d8' was found on the request parameter '_csrf' or header 'X-CSRF-TOKEN'.","path":"/api/users"}

Whatever you try to update, the same issue occurs.

I spent some time with this trying to resolve but haven't managed to progress.

Regards

**Comment by imilic1** [comment]
Just to add, record doesn't get saved in any case.

**Comment by jdubois** [comment]
Yes, your issue is with Hazelcast, we are currently working on it

**Comment by imilic1** [comment]
Thank you @jdubois

**Comment by jdubois** [comment]
Cleaned up the configuration in https://github.com/jhipster/generator-jhipster/commit/1084e4c2253a44cad597ab23774bf1f7dcc97e83 but the issue is still here.

What is strange is that:

- Login -> go to settings page -> save the settings -> you get the CSRF error
- Login -> go to monitoring page -> go to settings page -> save the settings -> everything is OK

Same issue with the logout, after that you can't login again unless you refresh the browser.

So from what I understand we have an old CSRF token in the session, only when Hazelcast is used: it looks like the Hazelcast filter isn't at the right place, as after one new request it works=;

**Comment by jdubois** [comment]
So after a lot of debugging:

For some reason, when using Hazelcast, the CSRF token which should be store in http session is null after the first REST request.

- If your second request is a GET, you will get a new token and everything will work
- If your second request is a POST, as your token is null, it will fail

I guess there is something in Hazelcast that makes the first request not having the correct token, but I can't find out why.

An ugly solution, which I am committing now as it solves other issues, is #2480 

- You will still have the 1st request failing
- But this will renew the CSRF token and re-do the request, and this time it will work

So basically that solves the issue, the only issue is that when your first request is a POST, you will have a failed request in your logs (but the end-user won't see it).

I'll keep this opened for some time, if we don't find a better solution then we will have to stay that way.

**Comment by imilic1** [comment]
Thank you @jdubois - this is really strange one. I also spent a lot of time debugging but no luck.

**Comment by imilic1** [comment]
@jdubois  - I applied now changes from #2480 and apparently the code added to auth.interceptor.js is causing infinite loop in browser, after 403 response is hit and CSRF token is not empty.

**Comment by jdubois** [comment]
Oops! I don't have this issue, can you tell me what you do, so I can reproduce?

**Comment by imilic1** [comment]
I replicated it on my new entity, I'll try to replicate on system application.

**Comment by imilic1** [comment]
OK, I have some additions to jHipster front end calls so in my case it actually gets in a loop when old call gets retried as it just keeps retrying as I am getting 403 for some other reason as well it seems.

**Comment by jdubois** [comment]
OK, but is this just limited to your changes? In that case I guess it's OK if we add this change

**Comment by imilic1** [comment]
@jdubois - yes, in your configuration it should never happen

**Comment by jdubois** [comment]
OK great let's do the next release then :-)
I'm closing this as this is "solved", but this is clearly not perfect, and as far as I'm concerned it's an Hazelcast issue (as the only difference is that we have Hazelcast set up, and as I followed very closely the Hazelcast official configuration)
